### PR TITLE
Pca tool

### PR DESCRIPTION
**Pull request by lorrainealisha75**

### CODE REVIEW

File: tools/sklearn/pca.xml [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    <description>with scikit-learn</description>
+    <macros>
+        <import>main_macros.xml</import>
+    </macros>
+    <expand macro="python_requirements"/>
+    <expand macro="macro_stdio"/>
+    <version_command>echo "@VERSION@"</version_command>
+    <command detect_errors="exit_code">
+        <![CDATA[
+        python3 '$__tool_directory__/pca.py'
```
Comment (simonbray) [inline tools/sklearn/pca.xml:11]:
```suggestion
        python '$__tool_directory__/pca.py'
```

File: tools/sklearn/pca.xml [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        -i '$infile'
+        -if '${infile.ext}'
```
Comment (simonbray) [inline tools/sklearn/pca.xml:13]: this is probably not necessary - I guess all inputs are tabular format?

File: tools/sklearn/pca.xml [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+            -n '$select_pca_type.number'
+        #end if
+        -t '$select_pca_type.select_pca_opts'
+        #if $select_pca_type.select_pca_opts == 'classical'           
+            -s '$select_pca_type.select_solver_type.svd_solver_opts'
+            #if $select_pca_type.select_solver_type.svd_solver_opts == 'arpack'
+                -tol $select_pca_type.select_solver_type.tolerance
+            #end if
+            $select_pca_type.whiten
+        #end if
```
Comment (simonbray) [inline tools/sklearn/pca.xml:29]: you can use `#elif` here

File: tools/sklearn/pca.xml [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+            -e '$select_pca_type.select_solver_type.eigen_solver_opts'
+            #if $select_pca_type.select_solver_type.eigen_solver_opts == 'arpack'
+                -tol $select_pca_type.select_solver_type.tolerance
+                #if $select_pca_type.select_solver_type.max_iter != ''
+                    -mi $select_pca_type.select_solver_type.max_iter
+                #end if
+            #end if
+        #end if
+        -o '$outfile'
+        -of '${outfile.ext}'
```
Comment (simonbray) [inline tools/sklearn/pca.xml:65]: again probably not necessary

File: tools/sklearn/pca.xml [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        </conditional>
+        <conditional name="select_pca_type">
+            <param name="select_pca_opts" type="select" label="Select PCA Type" help="Choose which flavour of PCA to use">
+                <option value="classical" selected="true">Classical PCA</option>
+                <option value="incremental">Incremental PCA</option>
+                <option value="kernel">Kernel PCA</option>
+            </param>
+            <when value="classical">
+                <param name="number" type="integer" optional="true" label="Number of components" help="Number of components to keep. If not set, all components are kept"/>
+                <param name="whiten" type="boolean" label="Whiten Components" truevalue="--whiten" falsevalue=""/>
```
Comment (simonbray) [inline tools/sklearn/pca.xml:96]: A help param would be useful here

ping @khanteymoori

File: tools/sklearn/pca.xml [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+                <conditional name="select_solver_type">
+                    <param name="svd_solver_opts" type="select" label="SVD Solver">
```
Comment (simonbray) [inline tools/sklearn/pca.xml:98]: again, help

File: tools/sklearn/pca.xml [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+                        <option value="poly">poly</option>
+                        <option value="rbf">rbf</option>
+                        <option value="sigmoid">sigmoid</option>
+                        <option value="cosine">cosine</option>
+                        <option value="precomputed">precomputed</option>
+                    </param>
+                    <when value="poly">
+                        <param name="gamma" type="float" optional="true" label="Gamma Value" help="Kernel coefficient for rbf, poly and sigmoid kernels. Ignored by other kernels"/>
+                        <param name="degree" type="integer" value="3" label="Degree of the polynomial" help="Degree for poly kernels. Ignored by other kernels"/>
+                        <param name="coef0" type="float" value="1.0" label="Coef0" help="Independent term in poly and sigmoid kernels. Ignored by other kernels"/>
```
Comment (simonbray) [inline tools/sklearn/pca.xml:128]: Is a better label than coef0 possible?

File: tools/sklearn/pca.xml [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+        <param name="header" type="boolean" label="Exclude Header" truevalue="--header" falsevalue="" help="If present, exclude the header row from the dataset"/>
+        <conditional name="select_cols">
+            <param name="select_cols_opts" type="select" label="Select Columns" help="Choose to select all columns, or exclude/include some">
+                <option value="all" selected="true">Select All</option>
+                <option value="include">Include</option>
+                <option value="exclude">Exclude</option>
+            </param>
+            <when value="include">
+                <param name="column_indices" type="text" value="0" label="Include Columns">
+                    <validator type="regex" message="Please specify only positive integers">(?:\d+\s)+\d+</validator>
```
Comment (simonbray) [inline tools/sklearn/pca.xml:79]: These validators seem complicated - it should be fine to just allow digits 0 to 9 and a space character, right? Something like: `<validator type="regex" message="Please specify only positive integers">^[0-9 ]+$</validator>`?

File: tools/sklearn/pca.xml [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+        </test>
+    </tests>
+    <help>
+        <![CDATA[
+.. class:: infomark
+**What it does**
+This tool takes a tabular input file (one data point per row, each column a variable) 
+and performs PCA using Singular Value Decomposition, returning an equally sized tabular 
+file with the first PC in the first column, second PC in the second column, etc.
+        ]]>
```
Comment (simonbray) [inline tools/sklearn/pca.xml:207]: This needs updating - maybe @khanteymoori can help with this?

File: tools/sklearn/pca.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
+import argparse
+import numpy as np
+from sklearn.decomposition import PCA, IncrementalPCA, KernelPCA
+
+
+def main():
+    parser = argparse.ArgumentParser(description='RDKit screen')
+    parser.add_argument('-i', '--infile',
+                        help="Input file")
+    parser.add_argument('-if', '--informat',
```
Comment (simonbray) [inline tools/sklearn/pca.py:10]: can remove this arg

**Comment by simonbray** [comment]
Thanks a lot @lorrainealisha75, this looks great. Will try and test it out locally over the next few days.

@khanteymoori, can you take a look as well?

File: tools/sklearn/pca.xml [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+    <tests>
+        <test>
+            <param name="infile" value="pca_input.dat" ftype="tabular"/>
+            <param name="select_cols_opts" value="include"  />
+            <param name="column_indices" value="0 1 3 5 7 4" />
+            <param name="number" value="5"/>
+            <param name="select_pca_opts" value="classical"/>
+            <param name="svd_solver_opts" value="arpack"/>
+            <param name="tolerance" value="0.4"/>
+            <output name="outfile" ftype='tabular' compare="sim_size" delta="1" file="pca_classical_output.dat"/>
```
Comment (simonbray) [inline tools/sklearn/pca.xml:164]: for each of these tests, can you remove `compare="sim_size"`? It should be possible to compare the outputs exactly

File: tools/sklearn/pca.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+    elif args.pca_type == 'kernel':
+        pca_params.update({'kernel': args.kernel, 'eigen_solver': args.eigen_solver})
+
+        if args.kernel == 'poly':
+            pca_params.update({'gamma': args.gamma, 'degree': args.degree, 'coef0': args.coef0})
+        elif args.kernel == 'rbf':
+            pca_params.update({'gamma': args.gamma})
+        elif args.kernel == 'sigmoid':
+            pca_params.update({'gamma': args.gamma, 'coef0': args.coef0})
```
Comment (anuprulez) [inline tools/sklearn/pca.py:84]: `'gamma': args.gamma` is common across all the compared conditions. It can be added to the line `pca_params.update({'kernel': args.kernel, 'eigen_solver': args.eigen_solver})` . If changed, then this condition `args.kernel == 'rbf'` would not be necessary.

**Comment by lorrainealisha75** [comment]
Closing this PR and creating a new one, as this one has unnecessary commits.

**Comment by simonbray** [comment]
I did some testing locally and realised it's not possible to select columns for inclusion/exclustion by column name (if there is a header). Probably the best way is to use the sklearn macros to specify input:
```
        <conditional name="input_options">
            <expand macro="data_input_options"/>
            <when value="tabular">
                <expand macro="samples_tabular" label1="Choose the training dataset containing features" multiple1="true" multiple2="false"/>
            </when>
            <when value="sparse">
                <expand macro="sparse_target"/>
            </when>
        </conditional>
```

Then this creates a json object which can be parsed in the python script.

File: tools/sklearn/test-data/pca_input.dat [hunk 13]
```diff
@@ -0,0 +1,300 @@
+9.579515262220434924e-01 3.726315038026500881e-02 1.192922318396479886e+00 2.208390853256675612e+00 1.545465138214503487e+00 6.587709093437858598e+00 2.676346016926679283e+00 2.002623182927643519e+00 1.960111203222609788e+00 1.244000891424809074e-01
```
Comment (simonbray) [inline tools/sklearn/test-data/pca_input.dat:1]: A small thing - can you also update this file to be tab separated? I think the sklearn input macros require it.
